### PR TITLE
delete Float8DynamicLinear

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #304

Summary:

We are standardizing on `Float8Linear` as the only float8 linear object:
1. the stack ending with
   https://github.com/pytorch-labs/float8_experimental/pull/300 moved
   all of the functionality of `Float8DynamicLinear` to `Float8Linear`.
   The default settings of `Float8Linear` are to use dynamic scaling.
2. this PR deletes `Float8DynamicLinear` from the codebase and patches
   the relevant callsites in fbsource.

Test Plan:

```
// all tests pass
./test_everything.sh

// also run all benchmarks and verify correctness
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D59342767](https://our.internmc.facebook.com/intern/diff/D59342767)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59342767).

File: benchmarks/bench_multi_gpu.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 import torch.multiprocessing as mp
 import torch.nn as nn
 import torch.utils.benchmark as benchmark
-from float8_experimental.float8_linear import Float8Linear
+from float8_experimental.float8_linear import Float8Linear, TensorScalingType
```
Comment (drisspg) [inline benchmarks/bench_multi_gpu.py:17]: How useful is this benchmark in general?

File: float8_experimental/float8_dynamic_utils.py [hunk 2]
```diff
@@ -53,64 +53,6 @@ def backward(ctx, gradY):
         return fp8_tensor, None
```
Comment (drisspg) [inline float8_experimental/float8_dynamic_utils.py:53]: do you think there is much value in putting this into dynamic utils vs into utils?

File: float8_experimental/float8_linear.py [hunk 3]
```diff
@@ -402,8 +402,8 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
 
     def scaling_repr(self):
         # add scaling settings without using too many characters
-        # example: "x:del,w:del,dldy:dyn"
-        return f"x:{self.scaling_type_x.short_str()},w:{self.scaling_type_w.short_str()},dldy:{self.scaling_type_dL_dY.short_str()}"
+        # example: "x_del_w_del_dldy_dyn"
+        return f"x_{self.scaling_type_x.short_str()}_w_{self.scaling_type_w.short_str()}_dldy_{self.scaling_type_dL_dY.short_str()}"
```
Comment (drisspg) [inline float8_experimental/float8_linear.py:406]: Why the change out of curiosity? I think the prior version might be a little more readable

File: test/test_compile.py [hunk 4]
```diff
@@ -48,8 +45,12 @@ def _test_compile_base(
     x = torch.randn(*x_shape, device="cuda", dtype=linear_dtype)
     m_ref = nn.Linear(16, 32, bias=True, device="cuda", dtype=linear_dtype)
 
-    m_fp8 = get_float8_linear(
-        linear_type, m_ref, emulate, scaling_type_x, scaling_type_w, scaling_type_dL_dY
+    m_fp8 = Float8Linear.from_float(
```
Comment (drisspg) [inline test/test_compile.py:48]: calling 'swap_..' on nn.Linear module returns a model out of place. I think its fine either way

File: test/test_dtensor.py [hunk 5]
```diff
@@ -171,37 +168,28 @@ def test_dtensor_fp8_autograd(mesh: DeviceMesh, size=16):
...
-        # same behavior as Float8Linear.
-        # TODO(future): add support for float8 all-gather with delayed scaling
-        # for activations and gradients.
-        extra_kwargs = {
-            "scaling_type_x": TensorScalingType.DYNAMIC,
-            "scaling_type_w": TensorScalingType.DYNAMIC,
-            "scaling_type_dL_dY": TensorScalingType.DYNAMIC,
-        }
+    # For now, just use Float8Linear with dynamic scaling, which is the
+    # same behavior as Float8Linear.
```
Comment (drisspg) [inline test/test_dtensor.py:175]: Float8Dynamic ? But also its probably to to just say, only supports dynamic scaling for all 3 tensors, x, w, dl_dY

File: test/test_fsdp2/test_fsdp2_common.py [hunk 6]
```diff
@@ -29,8 +27,7 @@ def check_parity_no_mp(
                 for param in model.parameters():
                     dist.all_reduce(param.grad)
                     param.grad.div_(dist.get_world_size())
-            if module_cls is Float8Linear:
-                sync_float8_amax_and_scale_history(model)
+            # TODO(future): add amax syncing once delayed scaling is supported
```
Comment (drisspg) [inline test/test_fsdp2/test_fsdp2_common.py:30]: was this just an unused code path?

File: test/test_fsdp2/test_fsdp2_eager.py [hunk 7]
```diff
@@ -76,16 +73,11 @@ def get_local_inp(self, dtype: torch.dtype = torch.float32):
...
-        self, module: nn.Module, use_float8_linear=False, **kwargs: Any
-    ) -> nn.Module:
-        if use_float8_linear:
-            kwargs["scaling_type_x"] = TensorScalingType.DYNAMIC
-            kwargs["scaling_type_w"] = TensorScalingType.DYNAMIC
-            kwargs["scaling_type_dL_dY"] = TensorScalingType.DYNAMIC
-            return swap_linear_with_float8_linear(module, Float8Linear, **kwargs)
-        else:
-            return swap_linear_with_float8_linear(module, Float8DynamicLinear, **kwargs)
+    def swap_linear_with_dynamic(self, module: nn.Module, **kwargs: Any) -> nn.Module:
```
Comment (drisspg) [inline test/test_fsdp2/test_fsdp2_eager.py:76]: can we just remove this since this is the default?

**Review by drisspg — APPROVED** [review]
Burn it with fire!🔥

File: test/test_fsdp2/test_fsdp2_common.py (hunk 6 above)
Comment (vkuzo) [reply test/test_fsdp2/test_fsdp2_common.py:30]: yes

File: benchmarks/bench_multi_gpu.py (hunk 1 above)
Comment (vkuzo) [reply benchmarks/bench_multi_gpu.py:17]: I haven't used it recently

File: float8_experimental/float8_dynamic_utils.py (hunk 2 above)
Comment (vkuzo) [reply float8_experimental/float8_dynamic_utils.py:53]: I think the git history is a bit cleaner this way which is why it made sense to me for this PR (minimizes changes to utils.py).  I'm not opposed to consolidate in a future PR, don't feel strongly one way or the other.

File: float8_experimental/float8_linear.py (hunk 3 above)
Comment (vkuzo) [reply float8_experimental/float8_linear.py:406]: I should have reverted this.  Will follow-up in a future PR if that's ok, to make landing this PR easier.

File: test/test_compile.py (hunk 4 above)
Comment (vkuzo) [reply test/test_compile.py:48]: I agree, we can make the tests use that if we want in a future PR.

File: test/test_dtensor.py (hunk 5 above)
Comment (vkuzo) [reply test/test_dtensor.py:175]: agreed, let me fix in a future PR to speed up landing this, since this is a minor point.

File: test/test_fsdp2/test_fsdp2_eager.py (hunk 7 above)
Comment (vkuzo) [reply test/test_fsdp2/test_fsdp2_eager.py:76]: agreed in principle, but ideally that would be a separate PR since it's only tangentially related
